### PR TITLE
Fully support Permit2 in Router and add function that can handle dynamic maxRedeem

### PR DESCRIPTION
**Pull request by haydenshively**

### CODE REVIEW

File: periphery/src/Router.sol
Comment (aneesh1) [inline periphery/src/Router.sol:76]: I might not have understood the `Permit2` docs correctly, so please correct me if I'm wrong. Clients of this function must call `ERC20(token).approve(Permit2)`. Is this `Permit2` Contract the same one that has been passed to the `Router`?

File: periphery/src/Router.sol
Comment (aneesh1) [inline periphery/src/Router.sol:100]: When might this function be used?

File: periphery/src/Router.sol
Comment (aneesh1) [inline periphery/src/Router.sol]: This looks like you're using `Permit` from [EIP-2612](https://github.com/dragonfly-xyz/useful-solidity-patterns/tree/main/patterns/permit2#permit-eip-2612-model) instead of `Permit2`. Is that correct? If so, the link provided mentions that this isn't supported by all major tokens:

> Since EIP-2612 is an extension of the ERC20 standard, this functionality is only possible on new (or upgradeable) tokens. So there are very few major tokens in the wild where this pattern actually works.

Should this also be using `Permit2`?

**Review by aneesh1 — COMMENTED** [review]
Read over some documentation [here](https://github.com/dragonfly-xyz/useful-solidity-patterns/tree/main/patterns/permit2#permit-eip-2612-model) prior to reviewing. I have some questions about how clients ought to use `repayWithPermit2` and a question about the `redeemWithChecks` function.

File: periphery/src/Router.sol
Comment (haydenshively) [reply periphery/src/Router.sol:76]: Correct.

File: periphery/src/Router.sol
Comment (haydenshively) [reply periphery/src/Router.sol:100]: Used to determine whether a user's `maxRedeem` number of shares is changing. If it is, we may use the helper function on `Router` to redeem instead of plain `lender.redeem(...)`

File: periphery/src/Router.sol
Comment (haydenshively) [reply periphery/src/Router.sol]: Yes, this is using EIP2612. The `Router` needs permission to burn the user's shares. Since that permission is governed by our `Lender` contract (which implements ERC20 and the EIP2612 extension) we don't need to worry about compatibility.

**Review by aneesh1 — APPROVED** [review]

**Comment by haydenshively** [comment]
@aneesh1 could you re-approve this please? Github seems not to have noticed that you approved it before, and won't let me merge.

**Review by aneesh1 — APPROVED** [review]

**Review by aneesh1 — APPROVED** [review]

**Comment by aneesh1** [comment]
It's still saying "Merging is blocked", but I have no idea why. As a workaround, maybe we request @IanWoodard's review and then he can approve to unblock?

**Review by IanWoodard — APPROVED** [review]
LGTM

**Pull request merged**
